blog/modules/ROOT/pages/using-types.adoc

```diff
@@ -0,0 +1,220 @@
+So let's have a look at how Pkl can go beyond most traditional types.
+
+== Beyond Strings
+
+`String` is a type we tend to see everywhere. This is especially true in configuration
+languages, where a `String` can be used as a "joker" type for many different values.
+
+=== Enumerating possibilities
+
+A typical usage of Strings is to define a finite set of values that your property can take:
+
+[source,pkl]
+----
+/// The environment where this service is running.
+///
+/// Valid possibilities are "dev", "qa" and "prod".
+env: String
+----
+
+This works, but has the problem that users of your template can provide invalid data, like `"foo"`.
+So let's do better:
+
+[source,pkl]
+----
+/// The environment where this service is running.
+env: "dev"|"qa"|"prod"
+----
+
+Now `env` will reject any invalid possibility. If you're using an IDE, like the
+link:https://github.com/apple/pkl-intellij[IntelliJ plugin], this has the added benefits
+of providing autocomplete and telling you when a given value is invalid.
+
+In Pkl, strings are not only values, they can also be types.
+
+=== Matching values
+
+Another typical use for strings is to allow values that have a specific *shape*:
+
+[source,pkl]
+----
+/// The IPv4 to connect
+serverIp: String
+----
+
+Again, we can do better to validate that the given ip is correct:
+
+[source,pkl]
+----
+/// The IPv4 to connect
+serverIp: String(matches(Regex(#"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}"#))) // <1>
+----
+<1> This regex is kept simple for illustration purposes, it still allows for invalid inputs.
+
+Strings can be constrained to match specific regular expressions.
+The new `serverIp` will only allow values of a specific shape, greatly reducing the possibility
+of invalid configs.
+
+=== More String goodies
+
+[source,pkl]
+----
+name: String(!isBlank) // <1>
+
+email: String(contains("@")) // <2>
+
+swiftFile: String(endsWith(".swift")) // <3>
+----
+<1> `name` cannot be empty or a blank String.
+<2> `email` must contain a `@` symbol.
+<3> A Swift file must have the correct extension.
+
+== Constraining numbers
+
+Pkl comes with some useful typealiases for numbers like `Int16`, `UInt8`, etc.
+
+[source,pkl]
+----
+serverPort: UInt16 // <1>
+----
+<1> Will only allow values between 0 and 65535, inclusive.
+
+Builtin typealiases are not the only way to handle such cases, though.
+Numbers can be constrained to any value.
+
+[source,pkl]
+----
+volume: Float(isBetween(0.0, 11.0)) // <1>
+
+multiplier: Int(isEven) // <2>
+----
+<1> `volume` can only be in the range between 0 and 11.
+<2> `multiplier` cannot be odd.
+
+== Beyond builtins
+
+Up to now, we only took a look at builtin functions, mostly functions of the type itself.
+We also limited our types to a single constraint. Pkl allows, though, for arbitrary functions as constraints,
+and we can provide multiple ones:
+
+[source,pkl]
+----
+local isPrime = (n: Int) -> IntSeq(2, n - 1).toList().every((x) -> n % x != 0)
```

Comment:
Nit: need to handle if `n` is less than 2

Reply:
Fixed.